portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -15,27 +15,73 @@
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
-@WebServlet("/data")
+@WebServlet("/comment-data")
 public class DataServlet extends HttpServlet {
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        response.setContentType("application/json");
+        ArrayList<Comment> comments = new ArrayList<>();
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        int lim = getCommentLimit(request);
+        int cnt = 0;
 
-  @Override
-  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-
-    ArrayList<String> messages = new ArrayList<>();
-    messages.add("msg1");
-    messages.add("msg2");
-    messages.add("msg3");
-    Gson gson = new Gson();
-    String json = gson.toJson(messages);
-    response.getWriter().println(json);
-  }
+        Query query = new Query("Comment").addSort("time", SortDirection.DESCENDING);
+        PreparedQuery results = datastore.prepare(query);
+        for (Entity entity : results.asIterable()) {
+            String username = (String)entity.getProperty("username");
+            String comment = (String)entity.getProperty("comment");
+            long time = (long)entity.getProperty("time");
+            comments.add(new Comment(username, comment, System.currentTimeMillis()-time));
+            cnt++;
+            if(cnt == lim) break;
+        }
+
+        String json = new Gson().toJson(comments);
+        response.getWriter().println(json);
+    }
+
+    @Override
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        Entity commentEntity = new Entity("Comment");
+        commentEntity.setProperty("username", getClientUsername(request));
+        commentEntity.setProperty("comment", getClientComment(request));
+        commentEntity.setProperty("time", System.currentTimeMillis());
+
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        datastore.put(commentEntity);
+
+        response.sendRedirect("/index.html");
+    }
+    
+    private String getClientUsername(HttpServletRequest request) {
+        return request.getParameter("username");
+    }
+
+    private String getClientComment(HttpServletRequest request) {
+        return request.getParameter("comment");
+    }
+
+    private int getCommentLimit(HttpServletRequest request) {
+        String s = request.getQueryString();
+        if(s == null) {
+            return 10;
+        } else {
+            return Integer.parseInt(s.substring(8,s.length()));
+        }
+    }
```

Comment:
Why not use request.getParameter?

Reply:
I thought I tried that and it didn't work, but now I see that it does work. Thanks for the suggestion!